[cmsp-679] PHP 8.3 compatibility checks

jazzsequence opened the pull request:
This PR adds PHP8.3 testing. 

As part of the investigation of this PR, I looked into using [PHPStan](https://github.com/phpstan/phpstan) (and the [PHPStan-WordPress](https://github.com/szepeviktor/phpstan-wordpress) extension) but found that those are more general purpose, code style checks that may or may not provide any insight into PHP 8+ compatibility.

It uses the [`phpcompatibility-action`](https://github.com/pantheon-systems/phpcompatibility-action/) to check PHPCompatibility as a new GH action. (PHPCompatibility-WP is part of Pantheon WP Coding Standards but PHPCompatibility does not support 8.x on the stable branch).

The version of PHP that is used when running the PHP Unit tests has also been updated to PHP 8.3 rather than 8.2. Anything that would be an issue with earlier versions of PHP 8.x would still be an issue in 8.3 so there's no benefit to running PHP Unit in both 8.2 _and_ 8.3.

The Behat tests are still running on a PHP 8.2 container. We would need a new build tools image in Quay that used PHP 8.3, so this change will need to wait until that image exists.

This PR also adds a new script for setting up PHPUnit locally which resolves the file not found error when it's looking for `wordpress-test-libs` in the wrong location.

jazzsequence committed:
add some local test helpers

jazzsequence committed:
let's see where the tests dir lives on ci

jazzsequence committed:
try another thing maybe?

jazzsequence committed:
add a new local phpunit install script and use that

jazzsequence committed:
use the right option

jazzsequence committed:
nocache isn't an option for composer...

jazzsequence committed:
remove the wp dir paths
we don't need them if we're setting TMPDIR

jazzsequence committed:
update the message about where we're installing tests

jazzsequence committed:
use newer versions of phpcompatibility for php8 checks

jazzsequence committed:
add php8 check and add to lint script

jazzsequence committed:
run the php8-check lint in php8.2 tests
the tests run composer phpcs but not composer lint. lint combines phpcs and php8-check so we'll run the php8-check manually

jazzsequence committed:
add new tests for 8.3

jazzsequence committed:
use dev-develop rather than dev-master

jazzsequence committed:
trash all that and use the GH action

jazzsequence committed:
restore line break that i inadvertently removed

jazzsequence committed:
just remove the 8.2 tests
we don't really need to test both things

pwtyler commented:
> The Behat tests are still running on a PHP 8.2 container. We would need a new build tools image in Quay that used PHP 8.3, so this change will need to wait until that image exists.

We would more importantly need an env running 8.3 on the fixture site, the env php version is more important than the build tools image 😆

pwtyler commented on bin/install-local-tests.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+set -e
+
+# Check if TMPDIR is set; if not, default to /tmp
```
nothing is checking TMPDIR

pwtyler commented on bin/install-local-tests.sh:
```diff
@@ -0,0 +1,21 @@
+
+# Run install-wp-tests.sh
+echo "Installing local tests into ${TMPDIR}"
+bash "$(dirname "$0")/install-wp-tests.sh" wordpress_test root '' 127.0.0.1 latest $SKIP_DB
```
should these have the option to override?

jazzsequence replied:
not sure what you mean. the point of the `$SKIP_DB` parameter is to optionally not install the database depending on whether you need it or not.

pwtyler replied:
i.e. on my local mariadb instance it got grumpy at me about using root/'', so for most local tests I have db creds configured with env variables for the given project with bashrc or .env. If this is a helper script for a human setting up testing, as opposed to CI, `wordpress_test root '' 127.0.0.1` might all want to be parameterized with defaults.

jazzsequence replied:
oh, you mean the db password, yeah, i could see that being parameterized.

jazzsequence replied:
I've made everything parameterized (or, ChatGPT did). I was mostly using this from the `composer test:install` and `composer test:install:withdb` scripts but a human can now run it manually with anything passed in as a flag.

jazzsequence committed:
update comment

jazzsequence committed:
run a phplint
composer lint is not run, only phpcs, and that's run in a non-php-specific environment, so splitting it out into a separate script and composer lint runs both

jazzsequence committed:
run the phplint script in both the 7.4 and 8.3 tests

jazzsequence committed:
allow any install-wp-tests parameter to be passed in
